tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py

```diff
@@ -0,0 +1,248 @@
+import logging
+
+import pytest
+
+from server import config
+from server.matchmaker import CombinedSearch, Search
+from server.matchmaker.algorithm.team_matchmaker import (
+    TeamMatchMaker,
+    UnevenTeamsException
+)
+
+
+@pytest.fixture(scope="module")
+def player_factory(player_factory):
+    player_id_counter = 0
+
+    def make(
+        mean: int = 1500,
+        deviation: int = 500,
+        ladder_games: int = config.NEWBIE_MIN_GAMES+1,
+        name=None
+    ):
+        nonlocal player_id_counter
+        """Make a player with the given ratings"""
+        player = player_factory(
+            ladder_rating=(mean, deviation),
+            ladder_games=ladder_games,
+            login=name,
+            lobby_connection_spec=None,
+            player_id=player_id_counter
+        )
+        player_id_counter += 1
+        return player
+    return make
+
+
+def make_searches(ratings, player_factory):
+    searches = []
+    for i, r in enumerate(ratings):
+        searches.append(Search([player_factory(r + 300, 100, name=f"p{i}")]))
+    return searches
+
+
+def test_team_matchmaker_performance(player_factory, bench, caplog):
+    # Disable debug logging for performance
+    caplog.set_level(logging.INFO)
+    num_searches = 200
+
+    searches = [Search([player_factory(1500, 500, ladder_games=1)]) for _ in range(num_searches)]
+
+    with bench:
+        TeamMatchMaker(4).find(searches)
+
+    assert bench.elapsed() < 0.5
+
+
+def test_team_matchmaker_algorithm(player_factory):
+    s = make_searches(
+        [1251, 1116, 1038, 1332, 1271, 1142, 1045, 1347, 1359, 1348, 1227, 1118, 1058, 1338, 1271, 1137, 1025],
+        player_factory)
+    c1 = CombinedSearch(*[s.pop(), s.pop()])
+    c2 = CombinedSearch(*[s.pop(), s.pop()])
+    c3 = CombinedSearch(*[s.pop(), s.pop()])
+    c4 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    s.append(c1)
+    s.append(c2)
+    s.append(c3)
+    s.append(c4)
+
+    matches = TeamMatchMaker(4).find(s)
+
+    assert len(matches) == 2
+    for match in matches:
```

Comment:
Assuming the algorithm is deterministic, it would probably be be better to `assert matches == [...]`. I guess it depends on how annoying it is to assign the searches to variables and stuff.

Reply:
I wanted to have the test more generic, because we don't care so much about a specific result, but about a good result. But maybe the better approach is to have the unit test very narrow and test the more general functionality with Hypothesis tests.

Reply:
Yea if you want to test properties then hypothesis is great. For me tests that check a specific example are also useful for documentation purposes, and it's a lot easier to read when you assert equality.